### DISCUSSION TITLE
Is it possible to specify a target file for "Move to a new file"?

### DISCUSSION THREAD

**Discussion by unphased** [typescript-language-server/typescript-language-server#850] [Q&A]
This code action is stupendously powerful. I am using neovim and I can use a range code action to refactor entire swathes of code. It: 

- generates a new source file in the same dir named after the first of the items i selected
- moves all selected items into this new source file
- refactors all import statements in the rest of my project that reference any of these items to import them out of the new source file
- in the original source file, cleans up any imports that were exclusively used by these items
- new source file accordingly also has those imports automatically inserted at the top

Needless to say this is going to save a lot of module import refactoring headaches and I am committed to leveraging automations like this as the new de facto workflow... the presence of this automation frankly makes me not want to use any other programming language...

My question is, how should I approach trying to make this particular code action more than only being useful for "breaking code apart into more files" workflows? I would like to be able to specify a target file to move some selected items into, rather than for a new filename to be automatically inferred. 

There is another plugin that i use which communicates to the LSP about filesystem actions performed, so I am able to use this code action and then rename the automatically created new file into a name that I want, this is very useful, however still does not address the issue that there's simply no way to move items freely between existing files. We can only freely break code into ever smaller code files and rename them. 

More specifically: 

- Does "Move to a new file" already accept a specified existing file to target?
- If not, it is clear that the building blocks needed to make such a functional extension to this action are already present. But, does the LSP protocol allow for us to provide input from the editor side of a target file path? What other challenges might exist here? GPT4 tells me: 
   > the LSP specification (as of my last update in April 2023) doesn't natively support sending arbitrary additional inputs for a standard textDocument/codeAction request. The parameters for this request are typically limited to: 
   > The text document's URI.
   > The range within the document.
   > The context, which includes diagnostics.

  I am inclined to believe this. If this is the case, then, what would be the most reasonable approach forward? Are there any extensions to the LSP protocol that can support this that we can start to implement? It's curious to me that the LSP protocol would support being able to send in a code position range but won't let us send in a target file path, or that there can't be a reply from the LSP server to ask for further information. I wouldn't mind in the short term doing "Move to a new file" first and triggering another code action from the existing target file, but it would also appear to be impossible to communicate to it all the details of that first action without resorting to sketchy out-of-band approaches, although in this case caching the path of the newly created file should in theory provide all of the necessary information.

**Comment by yioneko** [comment]
VSCode has a code action named "move to file", which implements the functionality you want.

It is currently impossible to implement this according to standard LSP, but possible through defining custom protocol extension.

**Reply by unphased** [reply]
Thanks i will give that a shot and see if i can get this working in vscode and then investigate how to go about it in neovim.

**Reply by parag776** [reply]
hey, i don't seem to specify a target file in Move to a new file, in vscode refactoring.

**Comment by danielpza** [comment]
Also tracked in this issue https://github.com/typescript-language-server/typescript-language-server/issues/863

**Comment by guillaumebrunerie** [comment]
This is now implemented in #987, but some custom implementation is required in the client (to show a file picker at the right time and do the right thing with the file path).

See [here](https://github.com/typescript-language-server/typescript-language-server/blob/master/docs/configuration.md#supportsmovetofilecodeaction-option) for what a client needs to implement and [here](https://github.com/emacs-lsp/lsp-mode/pull/4866) for an implementation of it in Emacs.

**Comment by ej-shafran** [comment]
For anyone coming here looking for a way to do this in Neovim, here's what I did:


```lua
local select = vim.ui.select
---@diagnostic disable-next-line: duplicate-set-field
vim.ui.select = function(items, opts, on_choice)
  if opts.kind == "codeaction" then
    local function handle_choice(item, idx)
      if
        item
        and item.action
        and item.action.kind == "refactor.move.file"
        and item.action.command
        and item.action.command.arguments
        and #item.action.command.arguments == 1
      then
        vim.ui.input({ prompt = "Destination file: ", completion = "file" }, function(input)
          if not input then
            return
          end

          item.action.command.arguments[1].interactiveRefactorArguments = {
            targetFile = vim.fn.fnamemodify(input, ":p"),
          }

          on_choice(item, idx)
        end)
        return
      end

      on_choice(item, idx)
    end
    return select(items, opts, handle_choice)
  end

  return select(items, opts, on_choice)
end
```

And in the `init_options` of the LSP client you have to add `supportsMoveToFileCodeAction = true`.
